File: 2022/solutions/day_04.py
parse = lambda x: [[*map(int, l.split(","))] for l in x.replace("-",",").splitlines()]

def part_1(data):
    return sum(1 for l in parse(data) if (l[0] <= l[2] and l[1] >= l[3]) or (l[2] <= l[0] and l[3] >= l[1]))

def part_2(data):
    return sum(1 for l in parse(data) if l[0] <= l[3] and l[1] >= l[2])


EX_1 = """\
2-4,6-8
2-3,4-5
5-7,7-9
2-8,3-7
6-6,4-6
2-6,4-8"""

def test():
    assert(part_1(EX_1) == 2)
    assert(part_2(EX_1) == 4)


if __name__ == "__main__":
    test()

File: 2017/src/solutions/day_04.rs
pub fn part_1(input: &str) -> String {
    input.lines().map(|l| {
        let mut words = l.split_whitespace().collect::<Vec<&str>>();
        let len = words.len();
        words.sort();
        words.dedup();
        (len == words.len()) as u32
    }).sum::<u32>().to_string()
}

pub fn part_2(input: &str) -> String {
    input.lines().map(|l| {
        let mut words = l.split_whitespace().map(|s| s.to_string()).collect::<Vec<String>>();
        let len = words.len();
        for i in 0..len {
            words[i] = {
                let mut chars = words[i].chars().collect::<Vec<char>>();
                chars.sort();
                chars.into_iter().collect::<String>()
            };
        }
        words.sort();
        words.dedup();
        (len == words.len()) as u32
    }).sum::<u32>().to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    const EX_0: &str = "\
aa bb cc dd ee
aa bb cc dd aa
aa bb cc dd aaa";

    const EX_1: &str = "\
abcde fghij
abcde xyz ecdab
a ab abc abd abf abj
iiii oiii ooii oooi oooo
oiii ioii iioi iiio";

    #[test]
    fn part_1_test() {
        assert_eq!(part_1(EX_0), "2");
    }

    #[test]
    fn part_2_test() {
        assert_eq!(part_2(EX_1), "3");
    }
}